android/java/org/chromium/chrome/browser/playlist/PlaylistHostActivity.java

```diff
@@ -0,0 +1,492 @@
+import com.brave.playlist.model.PlaylistEventModel;
+import com.brave.playlist.model.PlaylistItemModel;
+import com.brave.playlist.model.PlaylistItemOptionModel;
+import com.brave.playlist.model.PlaylistModel;
+import com.brave.playlist.model.PlaylistOptionsModel;
+import com.brave.playlist.util.ConstantUtils;
+import com.brave.playlist.util.PlaylistUtils;
+import com.brave.playlist.view.bottomsheet.MoveOrCopyToPlaylistBottomSheet;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import org.chromium.base.BraveFeatureList;
+import org.chromium.base.Log;
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.ChromeTabbedActivity;
+import org.chromium.chrome.browser.app.BraveActivity;
+import org.chromium.chrome.browser.flags.ChromeFeatureList;
+import org.chromium.chrome.browser.init.AsyncInitializationActivity;
+import org.chromium.chrome.browser.playlist.PlaylistServiceFactoryAndroid;
+import org.chromium.chrome.browser.playlist.settings.BravePlaylistPreferences;
+import org.chromium.chrome.browser.preferences.SharedPreferencesManager;
+import org.chromium.chrome.browser.util.TabUtils;
+import org.chromium.mojo.bindings.ConnectionErrorHandler;
+import org.chromium.mojo.system.MojoException;
+import org.chromium.playlist.mojom.Playlist;
+import org.chromium.playlist.mojom.PlaylistEvent;
+import org.chromium.playlist.mojom.PlaylistItem;
+import org.chromium.playlist.mojom.PlaylistService;
+import org.chromium.playlist.mojom.PlaylistServiceObserver;
+
+public class PlaylistHostActivity extends AsyncInitializationActivity
+        implements ConnectionErrorHandler, PlaylistOptionsListener, PlaylistServiceObserver {
+    private static final String TAG = "BravePlaylist";
+    private PlaylistService mPlaylistService;
+    private PlaylistViewModel mPlaylistViewModel;
+
+    @Override
+    public void onConnectionError(MojoException e) {
+        if (ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_PLAYLIST)
+                && SharedPreferencesManager.getInstance().readBoolean(
+                        BravePlaylistPreferences.PREF_ENABLE_PLAYLIST, true)) {
+            mPlaylistService = null;
+            initPlaylistService();
+        }
+    }
+
+    private void initPlaylistService() {
+        if (mPlaylistService != null) {
+            return;
+        }
+
+        mPlaylistService = PlaylistServiceFactoryAndroid.getInstance().getPlaylistService(this);
+    }
+
+    @Override
+    protected void triggerLayoutInflation() {
+        setContentView(R.layout.activity_playlist_host);
+        onInitialLayoutInflationComplete();
+    }
+
+    @Override
+    public void finishNativeInitialization() {
+        super.finishNativeInitialization();
+        if (ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_PLAYLIST)) {
+            initPlaylistService();
+            mPlaylistService.addObserver(this);
```

Comment:
we've discovered that there is a leak every time we use mojom observers. Mojom holds the object that is an observer forever without releasing it. I tried to dig but couldn't understand why, even though we close a connection, destroy and etc, observers are not garbage collected. Making Activities as observers leads to Activities never get destroyed and they sit in memory, every time you open an Activity, the new one object is created, with all the resources layouts and etc. I've made a workaround for Brave wallet in that PR https://github.com/brave/brave-core/pull/14725, on making separate objects for observers that are only observers, they still are not destroyed, but we can delete them in `onDestroy` on Activities and Activities with all resources are not leaked in that case. Please take a look on the above PR and make it similar in your case.